coordinator/src/main/scala/filodb.coordinator/QueryActor.scala

```diff
@@ -155,46 +170,62 @@ final class QueryActor(memStore: TimeSeriesStore,
             .onErrorHandle { t =>
               StreamQueryError(q.queryContext.queryId, querySession.queryStats, t)
             }.map { resp =>
-              FiloSchedulers.assertThreadName(QuerySchedName)
+            // Avoiding the assert when the InProcessPlanDispatcher is used. As it runs
+            // the query on the current/Actor thread instead of the scheduler
+            if (!q.dispatcher.isInstanceOf[InProcessPlanDispatcher]) {
+                FiloSchedulers.assertThreadName(QuerySchedName)
+              }
+              replyTo ! resp
               resp match {
-                case e: StreamQueryError => logQueryErrors(e.t)
+                case e: StreamQueryError =>
+                  logQueryErrors(e.t)
+                  queryErrors.increment()
+                  queryExecuteSpan.fail(e.t.getMessage)
+                  if (e.t.isInstanceOf[QueryTimeoutException] || e.t.isInstanceOf[AskTimeoutException]) throw e.t
                 case _ =>
               }
-              logger.debug(s"Sending response $resp to $replyTo")
-              replyTo ! resp
             }.guarantee(Task.eval {
               querySession.close()
               queryExecuteSpan.finish()
             }).completedL
-        } else {
+        } else { // TODO remove this block when query streaming is enabled and working well
           q.execute(memStore, querySession)(queryScheduler)
-            .map { res =>
-              // below prevents from calling FiloDB directly (without Query Service)
-              // UPDATE: 31/10/2022. Avoiding the assert when the InProcessPlanDispatcher is used. As it runs
+            .onErrorHandle { t =>
+              QueryError(q.queryContext.queryId, querySession.queryStats, t)
+            }.map { res =>
+              // Avoiding the assert when the InProcessPlanDispatcher is used. As it runs
               // the query on the current/Actor thread instead of the scheduler
               if (!q.dispatcher.isInstanceOf[InProcessPlanDispatcher]) {
                 FiloSchedulers.assertThreadName(QuerySchedName)
               }
               replyTo ! res
               res match {
-                case QueryResult(_, _, vectors, _, _, _) => resultVectors.record(vectors.length)
                 case e: QueryError =>
                   logQueryErrors(e.t)
                   queryErrors.increment()
                   queryExecuteSpan.fail(e.t.getMessage)
+                  if (e.t.isInstanceOf[QueryTimeoutException] || e.t.isInstanceOf[AskTimeoutException]) throw e.t
```

Comment:
Can this be a matcher, instead? Or are these the only two we will ever want to throw?

Reply:
Only these exceptions for now. I'll leave it as is and we can refactor if this list grows.